File: src/system.c
/*
  Smart search for upgrade paths
  Copyright (C) 2016 Andrew Leeming <andrew.leeming@codethink.co.uk> and
  Bob Mottram <bob.mottram@codethink.co.uk>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

#include "scalam.h"

/**
 * @brief Adds a program to a system from a directory containing repos
 * @param sys System object
 * @param repos_dir The directory which contains the repos
 * @param subdirectory The subdirectory of repos_dir currently being scanned
 * @param ctr Character position during scan of the returned directories
 * @returns zero on success
 */
int system_add_program_from_repo_directory(sc_system * sys, char * repos_dir,
                                           char * subdirectory, int ctr)
{
    char full_directory[SC_MAX_STRING];
    char current_checkout[SC_MAX_STRING];
    int line_number;

    if (ctr == 0)
        return 0;

    /* string terminator */
    subdirectory[ctr] = 0;

    /* the full path for the program repo */
    sprintf(full_directory,"%s/%s",repos_dir,subdirectory);

    /* set the name of the program */
    sprintf((&sys->program[sys->no_of_programs])->name, "%s", subdirectory);

    /* update the details for this program */
    if (program_repo_get_commits(full_directory,
                                 &sys->program[sys->no_of_programs]) != 0)
        return 1;

    /* check that there are some commits */
    if (sys->program[sys->no_of_programs].no_of_versions <= 0)
        return 2;

    /* get the current checkout commit */
    if (program_repo_get_current_checkout(full_directory, current_checkout) != 0)
        return 3;

    /* Get the array index from the checkout */
    line_number =
        get_line_number_from_string_in_file((&sys->program[sys->no_of_programs])->versions_file,
                                            (char*)current_checkout);
    if (line_number < 0)
        return 4;

    /* Invert the line number so that the last line in versions_file
       corresponds to version index zero. This just makes incrementing
       through versions more intuitive. */
    sys->program[sys->no_of_programs].version_index =
        sys->program[sys->no_of_programs].no_of_versions -
        line_number;

    /* increment the number of programs in the system */
    sys->no_of_programs++;

    return 0;
}

/**
 * @brief Create a dependency matrix for a system
 * @param sys System definition
 * @returns zero on success
 */
int system_create_dependency_matrix(sc_system * sys)
{
    int i;

    sys->dependency_probability =
        (double**)malloc(sizeof(double*)*SC_MAX_SYSTEM_SIZE);
    if (sys->dependency_probability ==  NULL)
        return 1;

    for (i = 0; i < SC_MAX_SYSTEM_SIZE; i++) {
        sys->dependency_probability[i] =
            (double*)malloc(sizeof(double)*SC_MAX_SYSTEM_SIZE);
        if (sys->dependency_probability[i] == NULL)
            return 2;
    }

    return 0;
}

/**
 * @brief Creates a system definition from a set of git repos within
 *        a given directory
 * @param sys System definition
 * @param repos_dir The directory where the repos exist
 * @returns zero on success
 */
int system_create_from_repos(sc_system * sys, char * repos_dir)
{
    char commandstr[SC_MAX_STRING];
    char directories[SC_MAX_STRING];
    char subdirectory[SC_MAX_STRING];
    char current_commit[SC_MAX_STRING];
    char head_commit[SC_MAX_STRING];
    int i, ctr, retval;

    /* find the subdirectories. This returns a string with directories listed
       such as:

       dir1/
       dir2/
       dir3/
    */
    sprintf(commandstr, "cd %s\nls -F | grep /", repos_dir);
    if (run_shell_command_with_output(commandstr, directories) != 0)
        return 1;
    if (directories[0] == 0)
        return 2;

    subdirectory[0] = 0;
    ctr = 0;

    /* clear the system so that it's initial state is consistent */
    memset((void*)sys, '\0', sizeof(sc_system));

    system_create_dependency_matrix(sys);

    /* scan the directories string extracting individual subdirectories */
    for (i = 0; i < strlen(directories); i++) {
        /* look for carriage returns */
        if ((directories[i] != '\n') && (i < strlen(directories)-1)) {
            /* read the directory name */
            if (directories[i] != '/')
                subdirectory[ctr++] = directories[i];
        }
        else {
            retval = system_add_program_from_repo_directory(sys, repos_dir,
                                                            (char*)subdirectory, ctr);
            if (retval != 0) {
                printf("system_add_program_from_repo_directory err %d", retval);
                return 3;
            }

            /* reset the character counter */
            ctr = 0;
        }
    }

    /* Handle any stragglers if we didn't encounter a final carriage
       return or / character */
    if (system_add_program_from_repo_directory(sys, repos_dir,
                                               (char*)subdirectory, ctr) != 0)
        return 4;

    return 0;
}

/**
 * @brief Tries to build each component of the system in a container.
 *        Scores for each program/genome are set based on build success
 *        and number of automated tests passed
 * @param population Population definition
 * @param pop_ix The index of the population
 * @returns float total score of this system
 */
float system_build(sc_population *population, int pop_ix)
{
    sc_program * program;
    sc_system_state * state;
    float score_sum=0;
    /* TODO */


    state=population->individual[pop_ix]->change;

    /* Cycle through all programs in system */
    int i;
    for(i=0; i<population->sys.no_of_programs; i++)
    {
        program=&population->sys.program[i];

        /* If marked to install, attempt it */
        if(state->installed[i])
        {
            /*
             * TODO
             *
             * Do some install
             * state->version_index[j]
             */
        }

        /* Do some scoring based on build/tests */
        population->individual[i]->score=0.0;    /* TODO */


        score_sum+=population->individual[i]->score;
    }


    return score_sum;
}

/** @brief Copies from one system object to another
 * @param destination Location to copy to
 * @param source Location to copy from
 * @returns zero on success
 */
int system_copy(sc_system * destination, sc_system * source)
{
    int i;

    destination->no_of_programs = source->no_of_programs;

    memcpy((void*)&destination->program, (void*)&source->program,
           sizeof(sc_program)*SC_MAX_SYSTEM_SIZE);

    if (system_create_dependency_matrix(destination) != 0)
        return 1;

    for (i = 0; i < SC_MAX_SYSTEM_SIZE; i++)
        memcpy((void*)destination->dependency_probability[i],
               (void*)source->dependency_probability[i],
               sizeof(double)*SC_MAX_SYSTEM_SIZE);

    return 0;
}

/**
 * @brief Compares two systems
 * @param sys1 The first system object
 * @param sys2 The second system object
 * @returns zero is the systems are the same
 */
int system_cmp(sc_system * sys1, sc_system * sys2)
{
    int i;

    if (sys1->no_of_programs != sys2->no_of_programs)
        return 1;

    if (memcmp((void*)&sys1->program, (void*)&sys2->program,
               sizeof(sc_program)*SC_MAX_SYSTEM_SIZE) != 0)
        return 2;

    for (i = 0; i < SC_MAX_SYSTEM_SIZE; i++)
        if (memcmp((void*)sys1->dependency_probability[i],
                   (void*)sys2->dependency_probability[i],
                   sizeof(double)*SC_MAX_SYSTEM_SIZE) != 0)
            return 3;

    return 0;
}

/**
 * @brief Frees memory for a system
 * @param sys System object
 */
void system_free(sc_system * sys)
{
    int i;

    for (i = 0; i < SC_MAX_SYSTEM_SIZE; i++)
        free(sys->dependency_probability[i]);

    free(sys->dependency_probability);
}

File: src/system.py
'''
  Smart search for upgrade paths
  Copyright (C) 2016 Andrew Leeming <andrew.leeming@codethink.co.uk> and
  Bob Mottram <bob.mottram@codethink.co.uk>

  This program is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

import copy
import os

from logger import logger
from utils import *
from program import *
from repo_type import *


try:
    import git
except ImportError:
    print("Error: GitPython libary missing. For installation instructions "
           "see: https://github.com/gitpython-developers/GitPython")
    sys.exit()

class System:
    def __init__(self, repo_path=None, definitions=None, programs=None):
        '''
        A system (collection of programs) can be created in at least 4 ways

        - (repo_path) Local directory containing git checkouts of programs
        - (definitions) Baserock stratam file
        - (programs) List of Program instances
        - Empty system - Programs to add later

        @param repo_path String Path to the directory that contains the system
        @param definitions String Path to definitions file that describes the system
        @param program Program[] List of programs in a system
        '''

        # Check that more than one entry hasn't been given
        if [repo_path, definitions, programs].count(None) < 2:
            logger.error(list2str([repo_path, definitions, programs]))
            raise TypeError(u"System expects ONLY a repo_path, definitions or programs, parameter")

        # Scan directory for programs (git)
        if repo_path is not None:
            logger.debug("Creating System from repo_path")
            self.programs=System.programsFromGitDirectory(repo_path)
            
            if len(self.programs)==0:
                logger.warning("No programs found on repo_path ({})".format(repo_path))
            
        # Parse definitions
        elif definitions is not None:
            #TODO
            logger.debug("Creating System from baserock definitions")
            self.programs=[]

        # List of program instances given
        elif programs is not None:
            #TODO check that this is a list and all members are Programs
            logger.debug("Creating System from list")
            self.programs=programs

        # Empty system, expected to use addProgram() after creation
        else:
            logger.debug("Creating an empty System")
            self.programs=[]

        logger.debug("Program list for this system: %s"%(list2str(self.programs)))

    @staticmethod
    def programsFromGitDirectory(repo_path):
        '''
        Get a list of Programs from a base directory that contains Git repos

        @param repo_path String Path to the directory where git checkouts are
        @return Program[]
        '''

        program_list=[]

        #Scan all the directories for git repos
        logger.debug("Searching inside %s"%repo_path)
        for d in list_dirs(repo_path):

            program_name=d

            program_path=os.path.join(repo_path, d)
            logger.debug("-%s : %s"%(program_name,program_path))

            try:
                program=Program(repo=DirType(program_path), name=program_name)
                program_list.append(program)
            except git.InvalidGitRepositoryError:
                #Not a git repo so skip over it
                logger.debug("Not a valid git repo")
                continue

        return program_list

    def addProgram(self, program):
        '''
        Adds a program to a system

        @param program Program The proram you wish to add to the System
        '''

        ##Type checking params
        if not isinstance(program, Program):
            raise TypeError(u"System.addProgram 'program' expects a Program instance")

        logger.debug("Adding %s to the system list"%program)
        self.programs.append(program)

    def getPrograms(self):
        '''
        Getter for the list of programs

        @return Program[]
        '''

        return self.programs
    
    def getRandomProgram(self,rand):
        '''
        Randomly selects a program in the system
        
        @return Program
        '''
        
        rand_index=rand%self.count()
        
        return self.programs[rand_index]
    
    def count(self):
        '''
        Counts the number of programs in the system
        
        @return int
        '''
        
        return len(self.programs)
 
    def dependencyMatrix(self):
        '''
        Create a dependency matrix for a system

        TODO
        '''
        pass

    def build(self):
        '''
        Tries to build each component of the system. Scores for each program/genome
        are set based on build success and number of automated tests passed

        TODO
        '''

        for prog in self.programs:
            #TODO do something
            pass


    def clone(self):
        '''
        Makes a copy of the current System instance

        @return System
        '''

        #TODO shallow or deep copy? Does it matter?

        # Make a shallow copy
        return copy.copy(self)
    
    def setLowestVersions(self):
        '''
        Sets all the programs at the lowest possible verison and checks them out
        '''
        
        logger.debug("Setting all programs at the earliest commit")
        
        for p in self.programs:
            p.setVersionIndex(0)
            p.checkout()
            
            logger.debug("{} set to commit {}".format(p.name, p.getCurrentVersion()))
        #end for
        
    
    def dump(self):
        '''
        Prints out (to the logger) as much information about the current
        system as it can. This is mainly used for debugging/logging purposes.
        '''
        
        spam_limit=10
    
        logger.debug("Dumping System object")
        #Print out all the programs
        for p in self.programs:
            logger.debug("- {}".format(p.getName()))
            spam_count=0
            #Print out all the versions
            for ver in p.getVersions():
                #Highlight/mark the current version
                if ver == p.getCurrentVersion():
                    logger.debug("  * {} *".format(ver))
                    spam_count=0
                elif spam_count<spam_limit:
                    logger.debug("  - {}".format(ver))
                
                spam_count+=1
                    
                if spam_count==spam_limit:
                    logger.debug("  - .. [limiting version spam output] ..")
        #end for

    @staticmethod		
    def latestVersion(system):		
        '''		
        Creates a goal based on the latest versions of software in the System		
                
        @param system (System) Set of programs to upgrade		
        @returns System instance		
        '''		
        
        ##Type checking params		
        if not isinstance(system, System):		
            raise TypeError(u"Goal 'system' expects a System instance")       		
        
        #Create an empty system and popululate one by one with highest version		
        #of each software		
        goal_system=System()  		
        for prog in system.getPrograms():		
            #Make a copy of the object and set version to the max		
            max_version=prog.clone()		
            latest_version=max_version.getCurrentHead()		
            max_version.setVersion(latest_version)		
                   
            #Add this to our goal system		
            goal_system.addProgram(max_version)		
                    
                    
        return goal_system
    
    def getMaxScore(self):
        '''
        Gets the total score of this system
        '''
        
        #TODO
        
        return 1.0
        
    
    def __eq__(self, sys):
        # If sys isn't a System, then it is clearly not equal
        if not isinstance(sys, System):
            return False

        return self.programs == sys.programs
